includes/parallax-image.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+class NARA_Parallax_Image {
+
+	/**
+	 * Setup hooks.
+	 */
+	public function __construct() {
+		add_shortcode( 'parallax_image', array( $this, 'parallax_image_display' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'parallax_image_enqueue_scripts' ) );
+		add_action( 'register_shortcode_ui', array( $this, 'parallax_image_shortcode_ui' ) );
+	}
+
+	/**
+	 * Display an image with data attributes to determine parallax behavior.
+	 *
+	 * @param $atts Shortcode attributes.
+	 *
+	 * @return string
+	 */
+	public function parallax_image_display( $atts ) {
+		$defaults = array(
+			'attachment' => '',
+			'size' => '',
+			'position' => 'absolute',
+			'animate_from' => '',
+			'speed' => '',
+			'offset' => '0',
+			'z_index' => '0',
+			'classes' => false,
+		);
+
+		$atts = shortcode_atts( $defaults, $atts );
+
+		// Bail if any attributes are empty.
+		if ( in_array( '', $atts, true ) ) {
+			return '';
+		}
+
+		$wrapper_classes = 'parallax-image-wrapper ' . esc_html( $atts['position'] );
+		$wrapper_style = 'z-index: ' . esc_html( $atts['z_index'] ) . ';';
+		$image_data = wp_get_attachment_image_src( esc_html( $atts['attachment'] ), esc_html( $atts['size'] ) );
+
+		if ( $atts['classes'] ) {
+			$wrapper_classes .= ' ' . esc_html( $atts['classes'] );
+		}
+
+		if ( 'relative' === $atts['position'] ) {
+			$wrapper_style .= ' height: ' . $image_data[2] . 'px;';
+		}
+
+		ob_start();
+
+		?>
+		<div class="<?php echo esc_attr( $wrapper_classes ); ?>"
+			 aria-hidden="true"
+			 style="<?php echo esc_attr( $wrapper_style ); ?>">
+			<image src="<?php echo esc_url( $image_data[0] ); ?>"
+				 width="<?php echo esc_attr( $image_data[1] ); ?>"
+				 height="<?php echo esc_attr( $image_data[2] ); ?>"
+				 data-speed="<?php echo esc_attr( $atts['speed'] ); ?>"
+				 data-animate-from="<?php echo esc_attr( $atts['animate_from'] ); ?>"
+				 data-offset="<?php echo esc_attr( $atts['offset'] ); ?>"
+				 class="parallax-image">
+		</div>
+		<?php
+
+		$content = ob_get_contents();
+		ob_end_clean();
+
+		return $content;
+	}
+
+	/**
+	 * Enqueue scripts used on the front end.
+	 */
+	public function parallax_image_enqueue_scripts() {
+		$post = get_post();
+
+		if ( isset( $post->post_content ) && has_shortcode( $post->post_content, 'parallax_image' ) ) {
```

Comment:
In cases like this, I'll often put the `wp_enqueue_script()` directly in the method that outputs the shortcode. It will then only be enqueued when the shortcode is used without having to use `has_shortcode()` on the content.

Does that seem reasonable?